server/src/Icepeak/Server/WebsocketServer/HandleClientMultiSubscription.hs

```diff
@@ -0,0 +1,140 @@
+import Control.Concurrent.MVar (MVar, takeMVar, tryTakeMVar, newMVar, readMVar)
+import Control.Exception (SomeAsyncException, SomeException, catch, finally, fromException, throwIO)
+import Control.Monad (forever)
+import Data.Aeson (Value)
+import Data.Foldable (for_)
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import Data.UUID (UUID)
+import System.Random (randomIO)
+
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as LBS
+import qualified Network.WebSockets as WS
+
+import Icepeak.Server.Core (Core, coreClients, withCoreMetrics, SubscriberState (SubscriberStateNew))
+import qualified Icepeak.Server.Metrics as Metrics
+import qualified Icepeak.Server.Subscription as Subscription
+
+-- * Client handling
+
+newUUID :: IO UUID
+newUUID = randomIO
+
+data ClientPayload
+  = Subscribe   { subscribePath :: [Text], token :: Text }
+  | UnSubscribe { unsubscribePath :: [Text] }
+
+instance Aeson.FromJSON ClientPayload where
+  parseJSON = undefined
+
+data ClientPayloadParsed
+  = ClientPayloadSuccessfulParse ClientPayload
+  | ClientPayloadFailedParse
+
+clientPayloadFromByteString :: LBS.ByteString -> ClientPayloadParsed
+clientPayloadFromByteString byteString = 
+  case Aeson.eitherDecode byteString of
+      (Left _jsonDecodeError) -> ClientPayloadFailedParse
+      (Right clientPayload)   -> ClientPayloadSuccessfulParse clientPayload
+
+clientPayloadFromDataMessage :: WS.DataMessage -> ClientPayloadParsed
+clientPayloadFromDataMessage (WS.Text encodedUtf8ByteString _mbDecodedCache) =
+  clientPayloadFromByteString encodedUtf8ByteString
+clientPayloadFromDataMessage (WS.Binary _) = ClientPayloadFailedParse
+  
+authorisePathSubscription :: [Text] -> Text -> IO Bool
+authorisePathSubscription = undefined
+
+handleClient :: WS.Connection -> Core -> IO ()
+handleClient conn core = do
+  uuid <- newUUID
+
+  isDirtyMVar <- newMVar ()
+  subscribedPathsMVar <- newMVar (HashMap.empty :: HashMap [Text] (MVar Value))
+
+  let
+    serverStateMVar = coreClients core
+
+    onConnect = withCoreMetrics core Metrics.incrementSubscribers
+
+    onDisconnect = do
+      paths <-  HashMap.keys <$> takeMVar subscribedPathsMVar
+      for_ paths
+        (\path -> modifyMVar_ serverStateMVar
+          (pure . Subscription.unsubscribe path uuid))
+      withCoreMetrics core Metrics.decrementSubscribers
+
+    onPayload (ClientPayloadSuccessfulParse (Subscribe newPath pathToken)) = do
+      isAuthorised <- authorisePathSubscription newPath pathToken
+
+      case isAuthorised of
+        False -> undefined
+        True -> do
+          pathValueMVar <- newEmptyMVar
+          modifyMVar_ subscribedPathsMVar
+            (pure . HashMap.insert newPath pathValueMVar)
+          modifyMVar_ serverStateMVar
+            (pure . Subscription.subscribe newPath uuid
+              (SubscriberStateNew (pathValueMVar, isDirtyMVar)))
```

Comment:
I presume the sub and unsub operations are idempotent. Because under the hood the subscription tree uses a `HashMap.delete` and `HashMap.insert`

Unless we want to be able to signal errors about duplicate subs and unsubs.